### ISSUE TITLE
redirect() and non-ascii characters

### ISSUE THREAD

**Issue by wombatonfire** [bottlepy/bottle#792]
Redirect to url with non-ascii characters causes an error:

Exception:

```
UnicodeEncodeError('ascii', u'http://localhost:9891/?status=1 \u044f\u043d\u0432\u0430\u0440\u044f', 32, 38, 'ordinal not in range(128)')
```

Traceback:

```
Traceback (most recent call last):
  File "C:\Users\user\Documents\Bitbucket\lab\bottle.py", line 862, in _handle
    return route.call(**args)
  File "C:\Users\user\Documents\Bitbucket\lab\bottle.py", line 1732, in wrapper
    rv = callback(*a, **ka)
  File "C:\Users\user\Documents\Bitbucket\lab\main.py", line 160, in register
    redirect(u'/?status=1 января') # temp
  File "C:\Users\user\Documents\Bitbucket\lab\bottle.py", line 2421, in redirect
    res.set_header('Location', urljoin(request.url, url))
  File "C:\Users\user\Documents\Bitbucket\lab\bottle.py", line 1539, in set_header
    self._headers[_hkey(name)] = [str(value)]
UnicodeEncodeError: 'ascii' codec can't encode characters in position 32-37: ordinal not in range(128)
```

**Comment by evuez** [comment]
Could you include your redirect code? Just tested on Python 2.7 and 3.4 and it works fine on both for me.

**Comment by wombatonfire** [comment]
Line that leads to the traceback above:

    redirect(u'/?status=1 января')

If I chenge it to:

    redirect('/?status=error')

I get no errors.

**Comment by evuez** [comment]
This should work in 0.13, but if you want to use 0.12.8 just update the [line 1539](https://github.com/bottlepy/bottle/blob/0.12.8/bottle.py#L1539) from

```python
self._headers[_hkey(name)] = [str(value)]
```

to 

```python
self._headers[_hkey(name)] = [value if isinstance(value, unicode)
                                            else str(value)]
```

**Comment by sharpaper** [comment]
This issue seems to be fixed in `0.13-dev` indeed, but it's still there in the last release, which is currently `0.12.16`. Since this is a major blocking bug, would it please be possible to add this fix to the next `0.12.17` release without waiting for the whole `0.13`? Thank you.

**Comment by defnull** [comment]
Okay, did I miss something? I tested it with Python 2.7 und 3.6 and cannot reproduce the error:

```
>>> try: bottle.redirect(u'/?status=1 января')
... except Exception as e: e
... 
Location: http://127.0.0.1/?status=1 января
Content-Type: text/html; charset=UTF-8
>>> try: bottle.redirect('/?status=1 января')
... except Exception as e: e
... 
Location: http://127.0.0.1/?status=1 января
Content-Type: text/html; charset=UTF-8
```

It seems to be fixed in Bottle 12.11 (current is 12.16).
